### PR TITLE
Process registration webform data

### PR DESCRIPTION
**Pull request by hamzaavvan**
Fixed errors: Promises must be handled appropriately from GoogleSheets.ts and SQR.functions.ts

### CODE REVIEW

**Comment by hamzaavvan** [comment]
[Process registration webform data](https://trello.com/c/ptt0cWCw/109-process-registration-webform-data)

File: functions/src/services/GoogleSheets.ts [hunk 1]
```diff
@@ -103,8 +103,8 @@ export default class GoogleSheets {
   }
 
   public async getRow(rowNumber: number): Promise<any> {
-    this.connect();
-    this.getHeaders();
+    await this.connect();
+    await this.getHeaders();
```
Comment (anantakrishna) [inline functions/src/services/GoogleSheets.ts]: This may be a good suggestion, however this is not related to your current task. See point 5 of https://trello.com/c/rkJKKl5d/107-git-conventions.

File: functions/src/modules/SQR.functions.ts [hunk 2]
```diff
@@ -20,52 +20,58 @@ export enum ISoundQualityReportSheet {
...
 
-/////////////////////////////////////////////////
-//          OnNewAllotment (DB create and update Trigger)
-//      1. Mark the files in the database --> { status: "Given" }
-//              Function --> updateFilesOnNewAllotment
-//
-//      2. Send an email to the assignee to notify them of the new allotments
-//              Function --> sendEmailOnNewAllotment
-/////////////////////////////////////////////////
-export const updateFilesOnNewAllotment = functions.database
```
Comment (anantakrishna) [inline functions/src/modules/SQR.functions.ts]: Why are you removing this?

File: functions/src/modules/SQR.functions.ts [hunk 3]
```diff
@@ -20,52 +20,58 @@ export enum ISoundQualityReportSheet {
...
-        timestampDone: null,
-      });
+/**
+ * Process registration of webform data
+ * 1. Registers user by transferring data into /users/ path
+ * 2. Check if users already exists
+ * 
+ * @function processRegistration()
+ */
+export const processRegistration = functions.database
```
Comment (anantakrishna) [inline functions/src/modules/SQR.functions.ts]: This function is not belonging to the SQR module.

File: functions/src/modules/SQR.functions.ts [hunk 4]
```diff
@@ -20,52 +20,58 @@ export enum ISoundQualityReportSheet {
...
+.ref('/webforms/registration/{registration_id}')
+.onCreate(async (snapshot, context) => {
+  const user = snapshot.val();
 
-      if (sqrError == undefined) {
-        // if Successful FILE Update, update the ALLOTMENT accordingly
+  const olduser = await db.ref("/users/").orderByChild("/emailAddress")
```
Comment (anantakrishna) [inline functions/src/modules/SQR.functions.ts]: Do we need an index on this field? If it doesn't exist then it should be added to the `database.rules.json`.

File: functions/src/modules/SQR.functions.ts [hunk 5]
```diff
@@ -20,52 +20,58 @@ export enum ISoundQualityReportSheet {
...
-        else
-          db.ref(`/sqr/allotments/${newDocKey}`).update({
-            filesAlloted: true,
-            sendNotificationEmail: true,
-          });
-      }
-    })
+  // If the user already exists stop the execution and log the error
+  if (olduser.val() !== undefined) {
+    console.log(`Error! User already exist with email: ${user.email_address}`)
```
Comment (anantakrishna) [inline functions/src/modules/SQR.functions.ts]: This should be a warning, and not just a log message. And this is not an error, this is a warning.

File: functions/src/modules/SQR.functions.ts [hunk 6]
```diff
@@ -20,52 +20,58 @@ export enum ISoundQualityReportSheet {
...
+    return false;
   }
-)
+
+  // if user email doesn't exists prepare the new_user Object
+  const new_user: any = {
+    timestamp: moment.format('x'),
+    emailAddress: user.email_address,
+    phoneNumber: user.contact_number,
+    isAvailableOnWhatsApp: user.i_am_available_on_whatsapp,
```
Comment (anantakrishna) [inline functions/src/modules/SQR.functions.ts]: Will “Yes” automatically transform into a boolean value?

File: functions/src/modules/SQR.functions.ts [hunk 7]
```diff
@@ -20,52 +20,58 @@ export enum ISoundQualityReportSheet {
...
+    location: user.country,
+    name: user.your_name,
+    services: (<any>Object).values(user.seva).join(", "),
```
Comment (anantakrishna) [inline functions/src/modules/SQR.functions.ts]: I cannot understand the logic of this transformation. I think that there is no transformation needed at all.

File: functions/src/modules/SQR.functions.ts [hunk 8]
```diff
@@ -20,52 +20,58 @@ export enum ISoundQualityReportSheet {
...
+    experience: user.experience,
+    influencer: user.where_did_u_hear_about_this_seva,
+    recommendedBy: user.recommended_by,
+    status: user.status,
```
Comment (anantakrishna) [inline functions/src/modules/SQR.functions.ts]: There is no `status` in the source data.

File: functions/src/modules/SQR.functions.ts [hunk 9]
```diff
@@ -20,52 +20,58 @@ export enum ISoundQualityReportSheet {
...
+  }
+
+  /**
+   * Assigning user roles to new_user 
+   * Get the first letters of service name and assign it to @var new_user's roles
+   */
+  user.services.map(name => name.split(' ').map(piece => piece[0]).join(''))
```
Comment (anantakrishna) [inline functions/src/modules/SQR.functions.ts]: I believe this can be done inline in the object construction above… However if it is not easy to do, no need wasting time.

The only thing is that there is no `user.services` property.

File: functions/src/modules/SQR.functions.ts [hunk 10]
```diff
@@ -20,52 +20,58 @@ export enum ISoundQualityReportSheet {
...
+ * Process registration of webform data
+ * 1. Registers user by transferring data into /users/ path
+ * 2. Check if users already exists
+ * 
+ * @function processRegistration()
+ */
+export const processRegistration = functions.database
+.ref('/webforms/registration/{registration_id}')
+.onCreate(async (snapshot, context) => {
+  const user = snapshot.val();
```
Comment (anantakrishna) [inline functions/src/modules/SQR.functions.ts]: To avoid confusion I would suggest
```suggestion
  const webform = snapshot.val();
```

File: functions/src/modules/SQR.functions.ts [hunk 11]
```diff
@@ -520,7 +526,8 @@ export const exportSubmissionsToSpreadsheet = functions.database
         token,
         unwantedParts,
       } = snapshot.val();
-      gsheets.appendRow({
+      
+      await gsheets.appendRow({
```
Comment (anantakrishna) [inline functions/src/modules/SQR.functions.ts]: Irrelevant change.

File: functions/src/modules/SQR.functions.ts (hunk 9 above)
Comment (hamzaavvan) [reply functions/src/modules/SQR.functions.ts]: You're right its not `user.services`. And it can't be done up in the Object that's why i did that separately.

File: functions/src/services/GoogleSheets.ts (hunk 1 above)
Comment (hamzaavvan) [reply functions/src/services/GoogleSheets.ts]: I was getting error during linting that's why i fixed that.

File: functions/src/modules/SQR.functions.ts (hunk 4 above)
Comment (hamzaavvan) [reply functions/src/modules/SQR.functions.ts]: @anantakrishna I did that so we can log the warning message.

Comment (anantakrishna) [reply functions/src/modules/SQR.functions.ts]: Your comment is irrelevant to my comment. Am I missing something?

File: functions/src/modules/SQR.functions.ts (hunk 7 above)
Comment (anantakrishna) [reply functions/src/modules/SQR.functions.ts]: Anyhow, this transformation is not needed here.

File: functions/src/modules/SQR.functions.ts (hunk 4 above)
Comment (hamzaavvan) [reply functions/src/modules/SQR.functions.ts]: I can't follow this: 
> Do we need an index on this field? 

Can you please elaborate?

File: functions/src/modules/SQR.functions.ts (hunk 7 above)
Comment (hamzaavvan) [reply functions/src/modules/SQR.functions.ts]: @anantakrishna See this is the error we get if we remove that type transformation.

image

Comment (anantakrishna) [reply functions/src/modules/SQR.functions.ts]: OK, I mistakenly was thinking that `webform.seva` is a string. No issues here. Thank you.

File: functions/src/modules/SQR.functions.ts (hunk 4 above)
Comment (anantakrishna) [reply functions/src/modules/SQR.functions.ts]: I've checked that we have an index on `emailAddress` property of user in `database.rules.json`, so no issues here. That's what I meant.

File: functions/src/modules/Users.functions.ts [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+  /**
+   * Assigning user roles to new_user 
+   * Get the first letters of service name and assign it to @var new_user's roles
+   */
+  webform.seva.map(name => name.split(' ').map(piece => piece[0]).join(''))
+              .map(role => new_user.roles[role] = true);
+  
+  // assigning user languages to new_user Object 
+  webform.languages.map(lang => new_user.languages[lang] = true);
```
Comment (anantakrishna) [inline functions/src/modules/Users.functions.ts]: I believe you should use [`reduce`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce) instead of `map` here. Map is not intended to have any side effect, and it is supposed to return a mapped array, which we don't need. Whereas you can use `reduce` to accumulate the roles in the object, and in this case you will be able to assign it directly to `newUser`.

File: functions/src/modules/Users.functions.ts [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+  
+  db.ref(`/users/`).push(new_user);
+
+  return 1;
```
Comment (anantakrishna) [inline functions/src/modules/Users.functions.ts]: Just curious, why 1? Why not true?

Comment (hamzaavvan) [reply functions/src/modules/Users.functions.ts]: I don't think it makes such impact, but i'll change it to `true` because its a good practice. Thanks!

image

File: functions/src/modules/Users.functions.ts [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+  // If the user already exists stop the execution and log the error
+  if (oldUser.val() !== undefined) {
+    console.warn(`Warning! User already exist with email: ${webform.email_address}`)
+    return false;
+  }
+
+  // if user email doesn't exists prepare the newUser Object
+  const newUser: any = {
+    timestamp: moment.format('x'),
```
Comment (anantakrishna) [inline functions/src/modules/Users.functions.ts]: Use `completed` from the webform data.

File: functions/src/modules/Users.functions.ts [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    emailAddress: webform.email_address,
+    phoneNumber: webform.contact_number,
+    isAvailableOnWhatsApp: webform.i_am_available_on_whatsapp === 1,
+    languages: webform.languages.reduce((map, lang) => {
```
Comment (anantakrishna) [inline functions/src/modules/Users.functions.ts]:
```suggestion
    languages: webform.languages.reduce((result, language) => {
```

File: functions/src/modules/Users.functions.ts [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+      map[lang] = true;
+      return map;
+    }, {}),
+    location: webform.country,
+    name: webform.your_name,
+    services: (<any>Object).values(webform.seva).join(", "),
+    experience: webform.experience,
+    influencer: webform.where_did_u_hear_about_this_seva,
+    recommendedBy: webform.recommended_by,
+    roles: webform.seva.reduce((map, role) => {
```
Comment (anantakrishna) [inline functions/src/modules/Users.functions.ts]:
```suggestion
    roles: webform.seva.reduce((result, service) => {
```

File: functions/src/modules/Users.functions.ts [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+      // Get the first letters of service name
+      const code = role.split(' ').map(name => name[0].join(''));
```
Comment (anantakrishna) [inline functions/src/modules/Users.functions.ts]: I had already pointed it out that you cannot take first letters to get a role name because the service name can be changed, and not in all cases it will work. There should be a dictionary. Like this:

Content Reporting → CR
Something Else → XX

Comment (hamzaavvan) [reply functions/src/modules/Users.functions.ts]: So you mean that in some cases role name would not be the first letters of words like this one `Sound Quality Reporting = SQR`

BTW this is the demonstration i.e result of the script:
image

Comment (anantakrishna) [reply functions/src/modules/Users.functions.ts]: Yes, I mean exactly this.

File: functions/src/modules/Users.functions.ts [hunk 18]
```diff
@@ -25,27 +24,40 @@ export const processRegistration = functions.database
     return false;
   }
 
+  const _services = {
```
Comment (anantakrishna) [inline functions/src/modules/Users.functions.ts]: Using an underscore is not a standard (AFAIK). Please rename to `serviceRoles`.
And why not to inverse the dictionary? It would simplify logic below. It would become just `result[serviceRoles[service]] = true;`. Isn't it? The only thing is to check for existence in a similar way as you do.
